engine/src/control.cpp

```diff
@@ -454,6 +454,59 @@ void MCControl::deselect()
 	}
 }
 
+void MCControl::removefromobjecttree(bool p_is_child)
+{
+    // IM-2012-05-16 [[ BZ 10212 ]] deleting the dragtarget control in response
+    // to a 'dragdrop' message would leave these globals pointing to the deleted
+    // object, leading to an infinite loop if the target was a field
+    if (MCdragdest == this)
+    {
+        MCdragdest = nil;
+        MCdropfield = nil;
+    }
+    
+    if (MCdragsource == this)
+        MCdragsource = nil;
+	
+	// Remove from the id cache before detaching from parent
+	MCObject::removefromobjecttree(p_is_child);
+    
+    switch (parent->gettype())
+    {
+        case CT_CARD:
+        {
+            if (!p_is_child)
+            {
+                MCCard *cptr = parent.GetAs<MCCard>();
+                cptr->removecontrol(this, False, True);
```

Comment:
This is a pretty sweet-looking refactor. 💯 

I suggest `parent.GetAs<MCCard>()->removecontrol(this, False, True);` but YMMV.